app/scripts/lib/rpc-method-middleware/handlers/wallet-createSession/handler.ts

```diff
@@ -117,55 +145,144 @@ async function walletCreateSessionHandler(
-        isChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+        isEvmChainIdSupported: networkClientExistsForChainId,
+        isEvmChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+        getNonEvmSupportedMethods: hooks.getNonEvmSupportedMethods,
+        isNonEvmScopeSupported: hooks.isNonEvmScopeSupported,
       },
     );
 
     const { supportedScopes: supportedOptionalScopes } = bucketScopes(
       optionalScopesWithSupportedMethodsAndNotifications,
       {
-        isChainIdSupported: networkClientExistsForChainId,
-        isChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+        isEvmChainIdSupported: networkClientExistsForChainId,
+        isEvmChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+        getNonEvmSupportedMethods: hooks.getNonEvmSupportedMethods,
+        isNonEvmScopeSupported: hooks.isNonEvmScopeSupported,
       },
     );
 
+    const allRequestedAccountAddresses = getAllAccountsFromScopesObjects([
+      supportedRequiredScopes,
+      supportedOptionalScopes,
+    ]);
+
+    const allSupportedRequestedCaipChainIds = getAllScopesFromScopesObjects([
+      supportedRequiredScopes,
+      supportedOptionalScopes,
+    ]);
+
     // Fetch EVM accounts from native wallet keyring
-    // These addresses are lowercased already
     const existingEvmAddresses = hooks
       .listAccounts()
       .map((account) => account.address);
-    const supportedEthAccounts = getEthAccounts({
-      requiredScopes: supportedRequiredScopes,
-      optionalScopes: supportedOptionalScopes,
-    })
-      .map((address) => address.toLowerCase() as Hex)
-      .filter((address) => existingEvmAddresses.includes(address));
+
+    const supportedRequestedAccountAddresses =
+      allRequestedAccountAddresses.filter(
+        (requestedAccountAddress: CaipAccountId) => {
+          const {
+            address,
+            chain: { namespace },
+            chainId: caipChainId,
+          } = parseCaipAccountId(requestedAccountAddress);
+          if (namespace === KnownCaipNamespace.Eip155) {
+            return existingEvmAddresses.some((existingEvmAddress) => {
+              return isEqualCaseInsensitive(address, existingEvmAddress);
+            });
+          }
+          const getNonEvmAccountAddressesForChainId =
+            hooks.getNonEvmAccountAddresses(caipChainId);
+
+          if (namespace === KnownCaipNamespace.Solana) {
+            return getNonEvmAccountAddressesForChainId.some(
+              (existingCaipAddress) => {
+                // solana addresses are case sensitive
+                return requestedAccountAddress === existingCaipAddress;
+              },
+            );
+          }
+          // TODO we may need to handle case sensitivity for other namespaces
+          // For now, we only support solana case sensitivity
```

Comment:
I don't understand the reasoning here. I feel fairly strongly that we should just use case sensitive comparison for everything that isn't EIP155. That is what the MultichainRouter is going to do too.

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/29887/commits/0739f4011195ed1854043ed013f56a589bb179e3